### PR TITLE
scp 2321 - Add computeTransaction for terms

### PR DESCRIPTION
**Pull request by hrajchert**
This PR is a pre-requisite to fix the functionality of the simulator replacing the contract with the continuation (instead of highlighting) and also for adding a blockly or flow simulator.
 
Pre-submit checklist:
- Branch
    - [X] Commit sequence broadly makes sense
    - [X] Key commits have useful messages
    - [X] Relevant tickets are mentioned in commit messages
    - [X] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [X] Self-reviewed the diff
    - [X] Useful pull request description
    - [X] Reviewer requested

Pre-merge checklist:
- [x] Someone approved it
- [x] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: marlowe-playground-client/src/Marlowe/Holes.purs [hunk 1]
```diff
@@ -943,3 +962,186 @@ termToValue _ = Nothing
...
+reduceContractStep :: S.Environment -> S.State -> Contract -> ReduceStepResult
+reduceContractStep env state contract = case contract of
+  Close -> NotReduced
+  Pay accId payee tok val cont -> Reduced cont
+  If obsTerm cont1 cont2 -> case fromTerm obsTerm of
+    Just obs ->
+      if S.evalObservation env state obs then
+        Reduced cont1
+      else
+        Reduced cont2
```
Comment (palas) [inline marlowe-playground-client/src/Marlowe/Holes.purs:1127]: Could move `Reduced` out of the `if` :)

File: marlowe-playground-client/src/Marlowe/Holes.purs [hunk 2]
```diff
@@ -943,3 +962,186 @@ termToValue _ = Nothing
...
+  semanticContract <- fromTerm startContract
+  let
+    semanticStep = S.reduceContractStep env startState semanticContract
+
+    termsStep = reduceContractStep env startState startContract
+  case semanticStep /\ termsStep of
+    (S.Reduced _ _ newState _ /\ Reduced newContract) -> reduceContractUntilQuiescent env newState newContract
+    -- This case is thought the Close contract, because in the semantic version, running a contract step
+    -- can do a payment (so the state is reduced), but in this version the close is never reduced
+    (S.Reduced _ _ newState _ /\ NotReduced) -> reduceContractUntilQuiescent env newState term
```
[the remark this answers is not in the document]
Comment (palas) [reply marlowe-playground-client/src/Marlowe/Holes.purs:1101]: Also, I think the top-level parenthesis around the pattern are not necessary here

**Review by palas — APPROVED** [review]
Clever :)
Looks good to me, just a couple of suggestions

File: marlowe-playground-client/src/Marlowe/Holes.purs (hunk 2 above)
[the remark this answers is not in the document]
Comment (hrajchert) [reply marlowe-playground-client/src/Marlowe/Holes.purs:1101]: good catch, applying

**Pull request merged**